source/connection-monitoring-and-pooling/tests/pool-checkout-returned-connection-maxConnecting.yml

```diff
@@ -0,0 +1,83 @@
+version: 1
+style: integration
+description: threads blocked by maxConnecting check out returned connections
+runOn:
+  -
+    # required for blockConnection in fail point
+    minServerVersion: "4.4.0"
+failPoint:
+  configureFailPoint: failCommand
+  mode: { times: 3 }
+  data:
+    failCommands: ["isMaster"]
+    closeConnection: false
+    blockConnection: true
+    blockTimeMS: 750
+poolOptions:
+  maxPoolSize: 10
+  waitQueueTimeoutMS: 5000
+operations:
+  # check out a connection and hold on to it.
+  - name: checkOut
+    label: conn0
+  # then start three threads that all attempt to check out. Two threads
+  # will fill maxConnecting, and the other should be waiting either for
+  # the other two to finish or for the main thread to check its connection
+  # back in.
+  - name: start
+    target: thread1
+  - name: checkOut
+    thread: thread1
+  - name: start
+    target: thread2
+  - name: checkOut
+    thread: thread2
+  - name: start
+    target: thread3
+  - name: checkOut
+    thread: thread3
+  # wait for all three to start checking out and a little longer
+  # for the establishments to begin.
+  - name: waitForEvent
+    event: ConnectionCheckOutStarted
+    count: 4
+  - name: wait
+    ms: 100
+  # check original connection back in, so the thread that isn't
+  # currently establishing will become unblocked. Then wait for
+  # all threads to complete.
+  - name: checkIn
+    connection: conn0
+  - name: waitForEvent
+    event: ConnectionCheckedOut
+    count: 4
+events:
+  # main thread checking out a Connection and holding it
+  - type: ConnectionCreated
+    address: 42
+    connectionId: 1
+  - type: ConnectionCheckedOut
+    address: 42
+  # two threads creating their Connections
+  - type: ConnectionCreated
+    address: 42
+  - type: ConnectionCreated
+    address: 42
+  # main thread checking its Connection back in
+  - type: ConnectionCheckedIn
+    connectionId: 1
+    address: 42
+  # remaining thread checking out the returned Connection
+  - type: ConnectionCheckedOut
+    connectionId: 1
```

Comment:
In Python this test fails here because we can't guarantee that `connectionId: 1` is checked out first. The issue is that using two semaphores is not fair (same problem discussed here: https://github.com/mongodb/specifications/pull/857#discussion_r495277704)

To address this I think we have to remove this line and omit connectionId.

Reply:
I left a similar comment on the Python PR about this--I think we should strive to pass the test as it is currently written if possible. It could really help out in situations where a ton of threads are queued up behind maxConnecting.

The following line in the version of the spec in this PR was meant to guard against this situation in most cases:

>Threads that are waiting [for pendingConnectionCount < maxConnecting or an available Connection] MUST be notified in order that they entered the WaitQueue."

To clarify my comment in the other thread: I don't think we need to absolutely guarantee fairness in every scenario, but I think we should try to eliminate unneeded waiting as much as possible and also ensure we avoid extremely unfair situations. In a scenario where we have a huge burst of operations that doesn't remain constant, we could have a lot of threads queued up for maxConnecting that aren't taking advantage of returned connections as quickly as they could be and subsequently receive their connection much later than newer threads. 

I now realize this is trickier to implement than just using another semaphore though, so if this ends up being really tricky / impossible to implement in Python, then I think we can just make it a SHOULD in the spec and edit this test as you suggest.

Reply:
Yeah I agree. I fixed this in the python implementation by using a condition variable instead of a semaphore.